pallets/loans-ref/src/lib.rs

```diff
@@ -0,0 +1,775 @@
+		NFTOwnerNotFound,
+		/// Emits when NFT owner doesn't match the expected owner
+		NotNFTOwner,
+		/// Emits when the applicant account is not the borrower of the loan
+		NotLoanBorrower,
+		/// Emits when the max number of active loans was reached
+		MaxActiveLoansReached,
+		/// Emits when the loan is bad specified and can not be created
+		CreateLoanError(CreateLoanError),
+		/// Emits when the loan can not be borrowed
+		BorrowLoanError(BorrowLoanError),
+		/// Emits when the loan can not be written off
+		WrittenOffError(WrittenOffError),
+		/// Emits when the loan can not be closed
+		CloseLoanError(CloseLoanError),
+	}
+
+	/// Error related to loan creation
+	#[derive(Encode, Decode, TypeInfo, PalletError)]
+	pub enum CreateLoanError {
+		/// Emits when valuation method is bad specified
+		InvalidValuationMethod,
+		/// Emits when repayment schedule is bad specified
+		InvalidRepaymentSchedule,
+	}
+
+	impl<T> From<CreateLoanError> for Error<T> {
+		fn from(error: CreateLoanError) -> Self {
+			Error::<T>::CreateLoanError(error)
+		}
+	}
+
+	/// Error related to loan borrowing
+	#[derive(Encode, Decode, TypeInfo, PalletError)]
+	pub enum BorrowLoanError {
+		/// Emits when the borrowed amount is more than the allowed amount
+		MaxAmountExceeded,
+		/// Emits when the loan can not be borrowed because the loan is written off
+		WrittenOffRestriction,
+		/// Emits when maturity has passed and borrower tried to borrow more
+		MaturityDatePassed,
+	}
+
+	impl<T> From<BorrowLoanError> for Error<T> {
+		fn from(error: BorrowLoanError) -> Self {
+			Error::<T>::BorrowLoanError(error)
+		}
+	}
+
+	/// Error related to loan borrowing
+	#[derive(Encode, Decode, TypeInfo, PalletError)]
+	pub enum WrittenOffError {
+		/// Emits when maturity has not passed tried to writ off
+		MaturityDateNotPassed,
+		/// Emits when a write off action tries to write off the more than the policy allows
+		LessThanPolicy,
+	}
+
+	impl<T> From<WrittenOffError> for Error<T> {
+		fn from(error: WrittenOffError) -> Self {
+			Error::<T>::WrittenOffError(error)
+		}
+	}
+
+	/// Error related to loan closing
+	#[derive(Encode, Decode, TypeInfo, PalletError)]
+	pub enum CloseLoanError {
+		/// Emits when close a loan that is not fully repaid
+		NotFullyRepaid,
+	}
+
+	impl<T> From<CloseLoanError> for Error<T> {
+		fn from(error: CloseLoanError) -> Self {
+			Error::<T>::CloseLoanError(error)
+		}
+	}
+
+	/// Creates a new loan against the collateral provided
+	///
+	/// The origin must be the owner of the collateral.
+	/// This collateral will be transferred to the existing pool.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(10_000)]
+		#[transactional]
+		pub fn create(
+			origin: OriginFor<T>,
+			pool_id: PoolIdOf<T>,
+			schedule: RepaymentSchedule,
+			collateral: AssetOf<T>,
+			collateral_value: T::Balance,
+			valuation_method: ValuationMethod<T::Rate>,
+			restrictions: LoanRestrictions<T::Rate>,
+			interest_rate_per_year: T::Rate,
```

Comment:
Having a `Rate` as `RateSec`/`RateYear` we can reduce this by allowing a `LoanInfo<T::RateYear>` parameter, and later transform it into a `LoanInfo<T::RateSec>` and avoid having both `LoanInfoInput` and `LoanInfo` which are basically the same.

Reply:
Blocked by https://github.com/centrifuge/centrifuge-chain/issues/1189